crates/database/src/orders_history.rs

```diff
@@ -0,0 +1,177 @@
+use {
+    crate::{jit_orders, orders, Address},
+    futures::stream::BoxStream,
+    sqlx::PgConnection,
+};
+
+pub fn user_orders<'a>(
+    ex: &'a mut PgConnection,
+    owner: &'a Address,
+    offset: i64,
+    limit: Option<i64>,
+) -> BoxStream<'a, Result<orders::FullOrder, sqlx::Error>> {
+    // As a future consideration for this query we could move from offset to an
+    // approach called keyset pagination where the offset is identified by "key"
+    // of the previous query. In our case that would be the lowest
+    // creation_timestamp. This way the database can start immediately at the
+    // offset through the index without enumerating the first N elements
+    // before as is the case with OFFSET.
+    // On the other hand that approach is less flexible so we will consider if we
+    // see that these queries are taking too long in practice.
+    #[rustfmt::skip]
+    const QUERY: &str = const_format::concatcp!(
+"(SELECT ", orders::ORDERS_SELECT,
+" FROM ", orders::ORDERS_FROM,
+" LEFT OUTER JOIN onchain_placed_orders onchain_o on onchain_o.uid = o.uid",
+" WHERE o.owner = $1",
+" ORDER BY creation_timestamp DESC LIMIT $2 + $3 ) ",
+" UNION ",
+" (SELECT ", orders::ORDERS_SELECT,
+" FROM ", orders::ORDERS_FROM,
+" LEFT OUTER JOIN onchain_placed_orders onchain_o on onchain_o.uid = o.uid",
+" WHERE onchain_o.sender = $1 ",
+" ORDER BY creation_timestamp DESC LIMIT $2 + $3 ) ",
+" UNION ",
+" (SELECT ", jit_orders::ORDERS_SELECT,
+" FROM jit_orders o",
+" WHERE o.owner = $1 AND NOT EXISTS (SELECT 1 FROM orders ord WHERE o.uid = ord.uid)",
+" ORDER BY creation_timestamp DESC LIMIT $2 + $3 ) ",
```

Comment:
Only this UNION with jit orders is a new code

Reply:
Honestly, given that we already fetch without offset from three tables (and only use the offset in the final union), I think we could definitely implement the merging and paginating in rust, but I won't insist.